### ISSUE TITLE
Changes due to Scouting, Crime, etc throw the following into strong focus

### ISSUE THREAD

**Issue by EldrickWT** [DFHack/dfhack#1279]
https://github.com/DFHack/dfhack/blob/8c8585d4eee5a9bad2b7aca58dde6ae032d03f00/library/modules/Units.cpp#L224-L227

When I commented this part out getVisibleName actually returned the visible identity. Cross verified in Dwarf Manipulator.

Currently, unit.name and the unit's hist_fig name match -for Criminals, Scouts, and other Agents.

I say 'into strong focus' and not more directly state to unilaterally comment it out, as I do not have a historical vampire to crosscheck. I do not know if their unit and hist_fig names match as well.

I've spent waaay too much effort on this [see https://github.com/EldrickWT/eld-dfhack-scripts/blob/master/toomuchidentity.lua in all it's spammy spammy glory.], and I'm starting to just see it on a dart board. As a target for assassination.

Thoughts?

**Comment by lethosor** [comment]
Looking back at this, I'm not entirely sure what you're saying. Are you suggesting removing that section as a fix? Do you happen to have a save handy?

**Comment by EldrickWT** [comment]
_**tl;dr section**_
**That is what I was suggesting.**

**As for a save** this was burneddi's save from May 17th
_**end tl;dr section**_

Sorry to ramble and muddle the issue, but I was drowning in frustration at the time. Especially that there wasn't a vampire in the fortress as well. Just the former spy running around on a prior identity because they'd _burned_ their actual identity due to a crime.

As for why the fix works... I *think* that way back when, identities got copied over the unit... change the identity and the unit got a copy. Worked great for vampires and werewolves.

Then Toady revisited falsifying your identity and changed it to not copy all the time. So the ident and the unit differed. Main DF knows to _always_ check the identities.

Is blasting the commented code out the ideal? I dunno. It worked when I did it with the spy. Don't know about vampires, etc.

**Comment by lethosor** [comment]
Oh, so maybe identities take precedence now. Should it check for identities, then histfig names, then unit names? It seems weird to skip histfigs to me, although it's possible that DF's nickname setting is like Units::setNickname, which sets both histfig and unit names when applicable.

**Comment by EldrickWT** [comment]
Well it was my fast and dirty fix, I didn't ask anyone competent to look it over until I suggested it here.

You probably have it right to not skip the hist fig. I was **massively** frustrated and couldn't see what the better precedence might be. When I finally nailed it down to the hist fig popping up after the identity and thus borking things I snapped it off like a rotten branch.

I was a bit too _salty_ at it to try to work it back in at a different precedence. I just knew main DF found identity to be key at this point, and was happy it worked on the spy.

**Comment by lethosor** [comment]
Save, for reference: https://drive.google.com/file/d/1bX5CQMqNsb_mjJOEOS4wBm4mGDLLs_cZ/view
I'll look into this asap

**Comment by lethosor** [comment]
Actually, that's using the name from a histfig ID associated with the identity, not with the unit. I think that could just be removed, then.

**Comment by cvuchener** [comment]
Is the unit list search using getVisibleName? Vampires need to be searched by their real name.

**Comment by lethosor** [comment]
getVisibleName is supposed to return the name visible in DF, so search should definitely be using it. I'm working on fixing it - with the histfig code removed, it handles this case as well as an apparent vampire from http://www.bay12games.com/dwarves/mantisbt/view.php?id=10530 (http://dffd.bay12games.com/file.php?id=13428).
